File: abc/abc276/e.cpp
#include <iostream>
#include <vector>
#include <set>
#include <tuple>

using namespace std;

int H, W;
vector<string> C;
vector<pair<int, int>> dir = { {-1, 0}, {0, 1}, {1, 0}, {0, -1} };

bool dfs(int i, int j, set<pair<int, int>>& visited, int depth, int start_i, int start_j) {
    visited.insert({i, j});

    for (const auto& d : dir) {
        int ni = i + d.first;
        int nj = j + d.second;

        if (0 <= ni && ni < H && 0 <= nj && nj < W) {
            if (ni == start_i && nj == start_j && depth >= 3) {
                return true;
            }
            if (visited.find({ni, nj}) == visited.end() && C[ni][nj] == '.') {
                if (dfs(ni, nj, visited, depth + 1, start_i, start_j)) {
                    return true;
                }
            }
        }
    }
    return false;
}

int main() {
    cin >> H >> W;
    C.resize(H);

    for (int i = 0; i < H; ++i) {
        cin >> C[i];
    }

    // Find the starting point 'S'
    int start_i = -1, start_j = -1;
    for (int i = 0; i < H; ++i) {
        for (int j = 0; j < W; ++j) {
            if (C[i][j] == 'S') {
                start_i = i;
                start_j = j;
                break;
            }
        }
        if (start_i != -1) break;
    }

    set<pair<int, int>> visited;
    bool result = dfs(start_i, start_j, visited, 0, start_i, start_j);

    cout << (result ? "Yes" : "No") << endl;

  }

File: abc/abc276/e.py
import sys

sys.setrecursionlimit(10**7)

H, W = map(int, input().split())
C = [input() for _ in range(H)]
dir = [(-1, 0), (0, 1), (1, 0), (0, -1)]


def dfs(i, j, visited, depth):
    visited.add((i, j))

    for di, dj in dir:
        ni, nj = i + di, j + dj
        if 0 <= ni < H and 0 <= nj < W:
            if (ni, nj) == (start_i, start_j) and depth >= 3:
                return True
            if (ni, nj) not in visited and C[ni][nj] == ".":
                res = dfs(ni, nj, visited, depth + 1)
                if res:
                    return True
    return False


start_i, start_j = next((i, j) for i in range(H) for j in range(W) if C[i][j] == "S")
print("Yes" if dfs(start_i, start_j, visited=set(), depth=0) else "No")
